utreexo-p2p-bip.md

```diff
@@ -1,236 +1,360 @@
+Both of these processes happen without any p2p messages by taking an already existing, synchronized archive full node and going through its stored block data.
 
-# Definitions
+Once an archive and bridge node have been established, CSNs download blocks and inclusion proofs to IBD and maintain sync with the bitcoin network. 
 
-Block Headers: The 80 byte block headers we all know and love.  The hash of the header is used as the unique identifier for the entire block.
+## Initial Block Download
 
-Block Summaries: A short (a few kilobytes) message about a block describing which UTXOs are spent in the block.  Tied to a specific block by including a block hash.
+![Current IBD](images/current-ibd.png)
 
-Block Proof:  A message containing hash based proof data which proves the validity of UTXOs being spent in the block.  Similar in size to a full block.  Also tied to a specific block by including a block hash.
+Current IBD is done by a headers-first block download, which downloads all the Bitcoin block headers, verifies that they connect and start downloading the actual block data for the headers.
 
-# Overview
+![Utreexo node IBD](images/utreexo-node-ibd.png)
 
-## Pre-P2P: Bridge Building
+Utreexo nodes will also perform the headers download but they also require the inclusion proof along with the block data.
+Hence a Utreexo node will send a `getutreexoproof` message along with the `getdata` message for a given block.
+This flow is the simpliest change and allows a Utreexo node to validate and perform IBD but this method does require downloading about 2 times compared to the current nodes as the inclusion proof for a block is roughly the same size as the block itself.
 
-When introducing utreexo into an existing network, there are 2 thing needed before CSNs can operate:  First, archive nodes need to build proofs for old blocks to serve during IBD, and second, nodes need to build and maintain the UTXO merkle forest, and an index of outpoints to leaves of that forest, so that they can build proofs for new transactions.  Both of these processes happen without any p2p messages by taking an already existing, synchronized archive full node and going through its stored block data.
+![Bandwidth efficient Utreexo node IBD](images/bandwidth-efficient-utreexo-node-ibd.png)
 
-Once an archive and bridge node have been established, CSNs use utreexo to IBD and maintain sync with the bitcoin network. 
+For utreexo nodes with memory to spare, we introduce a `ttl` messsage that will have a time-to-live value for each of the outputs in a given block.
+With these ttl values, a node receiving the `ttl` message will be able to determine which output to cache with the clairvoyant algorithm[^1] which allows the ibd-ing node to reduce the bandwidth required in syncing the node in the most efficient way possible.
```

Comment:
I think the same thing with paragraphs happened here

Reply:
I meant it as a big paragraph but maybe it's more readable if it's separated?